pkg/discovery-api/discovery_api.go

```diff
@@ -0,0 +1,180 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright 2020 Fairwinds
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+package discoveryapi
+
+import (
+	"context"
+	"encoding/json"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+
+	"github.com/fairwindsops/pluto/v5/pkg/api"
+)
+
+// DiscoveryClient is the
+type DiscoveryClient struct {
+	ClientSet       dynamic.Interface
+	restConfig      *rest.Config
+	DiscoveryClient discovery.DiscoveryInterface
+	Instance        *api.Instance
+}
+
+// NewDiscoveryClient returns a new struct with config portions complete.
+func NewDiscoveryClient(instance *api.Instance) *DiscoveryClient {
+	cl := &DiscoveryClient{
+		Instance: instance,
+	}
+
+	var err error
+	cl.restConfig, err = NewRestClientConfig(rest.InClusterConfig)
+	if err != nil {
+		return nil
+	}
+
+	if cl.DiscoveryClient, err = discovery.NewDiscoveryClientForConfig(cl.restConfig); err != nil {
+		return nil
+	}
+
+	cl.ClientSet, err = dynamic.NewForConfig(cl.restConfig)
+	if err != nil {
+		return nil
+	}
+	return cl
+}
+
+func NewRestClientConfig(inClusterFn func() (*rest.Config, error)) (*rest.Config, error) {
+
+	if restConfig, err := inClusterFn(); err == nil {
+		return restConfig, nil
+	}
+
+	pathOptions := clientcmd.NewDefaultPathOptions()
+
+	config, err := pathOptions.GetStartingConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	configOverrides := clientcmd.ConfigOverrides{}
+
+	clientConfig := clientcmd.NewDefaultClientConfig(*config, &configOverrides)
+	restConfig, err := clientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return restConfig, nil
+}
+
+// GetApiResources discovers the api-resources for a cluster
+func (cl *DiscoveryClient) GetApiResources() error {
+
+	resourcelist, err := cl.DiscoveryClient.ServerPreferredResources()
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return err
+		}
+		if apierrors.IsForbidden(err) {
+			klog.V(2).Infof("Failed to list objects for Name discovery. Permission denied! Please check if you have the proper authorization")
```

Comment:
Should we return this as an error? If we can't get the list of api-resources, there's no way to continue scanning in the cluster.

Reply:
Corrected.